File: stripe/src/main/java/com/stripe/android/Stripe3ds2CompletionStarter.java
package com.stripe.android;

import android.os.Bundle;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.stripe.android.model.StripeIntent;
import com.stripe.android.utils.ObjectUtils;
import com.stripe.android.view.AuthActivityStarter;
import com.stripe.android.view.PaymentRelayActivity;
import com.stripe.android.view.StripeIntentResultExtras;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

class Stripe3ds2CompletionStarter
        implements AuthActivityStarter<Stripe3ds2CompletionStarter.StartData> {
    @NonNull private final Host mHost;
    private final int mRequestCode;

    Stripe3ds2CompletionStarter(@NonNull Host host, int requestCode) {
        mHost = host;
        mRequestCode = requestCode;
    }

    @Override
    public void start(@NonNull StartData data) {
        final Bundle extras = new Bundle();
        extras.putString(StripeIntentResultExtras.CLIENT_SECRET,
                data.mStripeIntent.getClientSecret());
        extras.putInt(StripeIntentResultExtras.FLOW_OUTCOME, data.getOutcome());
        mHost.startActivityForResult(PaymentRelayActivity.class, extras, mRequestCode);
    }

    @IntDef({ChallengeFlowOutcome.COMPLETE_SUCCESSFUL, ChallengeFlowOutcome.COMPLETE_UNSUCCESSFUL,
            ChallengeFlowOutcome.CANCEL, ChallengeFlowOutcome.TIMEOUT,
            ChallengeFlowOutcome.PROTOCOL_ERROR, ChallengeFlowOutcome.RUNTIME_ERROR})
    @Retention(RetentionPolicy.SOURCE)
    @interface ChallengeFlowOutcome {
        int COMPLETE_SUCCESSFUL = 0;
        int COMPLETE_UNSUCCESSFUL = 1;
        int CANCEL = 2;
        int TIMEOUT = 3;
        int PROTOCOL_ERROR = 4;
        int RUNTIME_ERROR = 5;
    }

    static class StartData {
        @NonNull private final StripeIntent mStripeIntent;
        @ChallengeFlowOutcome private final int mChallengeFlowOutcome;

        StartData(@NonNull StripeIntent stripeIntent,
                  @ChallengeFlowOutcome int challengeFlowOutcome) {
            mStripeIntent = stripeIntent;
            mChallengeFlowOutcome = challengeFlowOutcome;
        }

        @StripeIntentResult.Outcome
        private int getOutcome() {
            if (mChallengeFlowOutcome == ChallengeFlowOutcome.COMPLETE_SUCCESSFUL) {
                return StripeIntentResult.Outcome.SUCCEEDED;
            } else if (mChallengeFlowOutcome == ChallengeFlowOutcome.COMPLETE_UNSUCCESSFUL) {
                return StripeIntentResult.Outcome.FAILED;
            } else if (mChallengeFlowOutcome == ChallengeFlowOutcome.CANCEL) {
                return StripeIntentResult.Outcome.CANCELED;
            } else if (mChallengeFlowOutcome == ChallengeFlowOutcome.TIMEOUT) {
                return StripeIntentResult.Outcome.TIMEDOUT;
            } else {
                return StripeIntentResult.Outcome.FAILED;
            }
        }

        @Override
        public int hashCode() {
            return ObjectUtils.hash(mStripeIntent, mChallengeFlowOutcome);
        }

        @Override
        public boolean equals(@Nullable Object obj) {
            return super.equals(obj) || (obj instanceof StartData && typedEquals((StartData) obj));
        }

        private boolean typedEquals(@NonNull StartData startData) {
            return ObjectUtils.equals(mStripeIntent, startData.mStripeIntent) &&
                    ObjectUtils.equals(mChallengeFlowOutcome, startData.mChallengeFlowOutcome);
        }
    }
}

File: stripe/src/main/java/com/stripe/android/Stripe3ds2CompletionStarter.kt
package com.stripe.android

import android.os.Bundle
import androidx.annotation.IntDef
import com.stripe.android.model.StripeIntent
import com.stripe.android.view.AuthActivityStarter
import com.stripe.android.view.PaymentRelayActivity
import com.stripe.android.view.StripeIntentResultExtras

internal class Stripe3ds2CompletionStarter(
    private val host: AuthActivityStarter.Host,
    private val requestCode: Int
) : AuthActivityStarter<Stripe3ds2CompletionStarter.Args> {

    override fun start(args: Args) {
        val extras = Bundle()
        extras.putString(StripeIntentResultExtras.CLIENT_SECRET,
            args.stripeIntent.clientSecret)
        extras.putInt(StripeIntentResultExtras.FLOW_OUTCOME, args.outcome)
        host.startActivityForResult(PaymentRelayActivity::class.java, extras, requestCode)
    }

    @IntDef(ChallengeFlowOutcome.COMPLETE_SUCCESSFUL, ChallengeFlowOutcome.COMPLETE_UNSUCCESSFUL,
        ChallengeFlowOutcome.CANCEL, ChallengeFlowOutcome.TIMEOUT,
        ChallengeFlowOutcome.PROTOCOL_ERROR, ChallengeFlowOutcome.RUNTIME_ERROR)
    @Retention(AnnotationRetention.SOURCE)
    internal annotation class ChallengeFlowOutcome {
        companion object {
            internal const val COMPLETE_SUCCESSFUL = 0
            internal const val COMPLETE_UNSUCCESSFUL = 1
            internal const val CANCEL = 2
            internal const val TIMEOUT = 3
            internal const val PROTOCOL_ERROR = 4
            internal const val RUNTIME_ERROR = 5
        }
    }

    internal data class Args internal constructor(
        val stripeIntent: StripeIntent,
        @param:ChallengeFlowOutcome @field:ChallengeFlowOutcome
        private val challengeFlowOutcome: Int
    ) {
        internal val outcome: Int
            @StripeIntentResult.Outcome
            @JvmSynthetic
            get() = when (challengeFlowOutcome) {
                ChallengeFlowOutcome.COMPLETE_SUCCESSFUL -> StripeIntentResult.Outcome.SUCCEEDED
                ChallengeFlowOutcome.COMPLETE_UNSUCCESSFUL -> StripeIntentResult.Outcome.FAILED
                ChallengeFlowOutcome.CANCEL -> StripeIntentResult.Outcome.CANCELED
                ChallengeFlowOutcome.TIMEOUT -> StripeIntentResult.Outcome.TIMEDOUT
                else -> StripeIntentResult.Outcome.FAILED
            }
    }
}
